### ISSUE TITLE
Prefetching of Tuples from Castor

### ISSUE THREAD

**Issue by sbckr** [carbynestack/ephemeral#6]
To perform secure computations, Ephemeral has to fetch tuples from Castor, which are then consumed throughout the execution of the MPC program.
However, Ephemeral is currently deployed with pre-generated tuples as streaming tuples from Castor to Ephemeral is not yet available. To increase security throughout the computation:
- Ephemeral must use the Castor API to fetch tuples and stream them into the MPC program execution
- Pre-generated tuples should be removed from the Ephemeral Docker Image

**Comment by sbckr** [comment]
Started feature request for streaming tuples in MP-SPDZ (data61/MP-SPDZ#418) as feature is not yet available.

**Comment by kindlich** [comment]
Feature to read from named pipes is now part of MP-SPDZ.
We need to upgrade MP-SPDZ in order to use it (WIP).

The general file-name from which MP-SPDZ will read is (when using a Prime Field)
```
Player-Data/<PartyCount>-p-<Prime-bit-length>/<TupleType>-p-P<PartyID>-T<ThreadID>

# e.g. for first Party, First Thread, reading Triples
Player-Data/3-p-128/Triples-p-P0-T0
```

Also, it should have Castor's Parallel Tuple Reservation working, because
- Multiple MP-SPDZ Instances could run in parallel
- A single MP-SPDZ program can use multiple tuple types of Tuples which could be fetched in parallel
- A single MP-SPDZ program can use multiple Threads, which require tuples

An idea on how it could run:
1) Ephemeral (?) creates FIFOs (also called "named pipes" in some systems)
2) Ephemeral starts background process
      2.1) Process starts one thread per Tuple Type and Thread -> Each Thread will write to one FIFO
      2.2) Each thread starts "writing" the File Header to its respective FIFO and thereby block until MP-SPDZ requires the tuples
      2.3) Once a thread's write is done (MP-SPDZ has read the header/tuple), the thread can create a request to castor to request/download X tuples of its type
      2.4) Thread writes Tuples to pipe and blocks until MP-SPDZ reads all of them
      2.5) Repeat from 2.3 onwards once all tuples have been read by MP-SPDZ
3) MP-SPDZ program returns
4) Ephemeral stops the background process


Open Questions:
- How many tuples are downloaded at once (configurable?) ?
   -> Prefetching more tuples means less requests to castor -> faster(?), but some tuples may be requested but never used
- How do we know how many the MP-SPDZ Program needs (can this be inferred from compilation?)
  -> We need one FIFO and Thread per (Tuple Type, MP-SPDZ Thread) combination
- How to make sure tuple fetch requests on all parties use the same Request ID
   -> Use named UUIDs consisting of...  
       - Thread number
       - Autoincrementing request number per Thread
       - Game ID
- Would such a multi-thread approach create special requirements to the pod running ephemeral (like how many vCores it has or so?)

**Comment by strieflin** [comment]
> How many tuples are downloaded at once (configurable?) ?
-> Prefetching more tuples means less requests to castor -> faster(?), but some tuples may be requested but never used

We could also think about giving tuples back to castor that have not been used. Would consider this as an optimization, however.

> How do we know how many the MP-SPDZ Program needs (can this be inferred from compilation?)
-> We need one FIFO and Thread per (Tuple Type, MP-SPDZ Thread) combination

We don't. Compiler spits out an estimation of required tuples only and can be completely wrong sometimes as of my understanding. This means, we need to dynamically handle the tuple fetching.

> How to make sure tuple fetch requests on all parties use the same Request ID
> -> Use named UUIDs consisting of...
> - Thread number
> - Autoincrementing request number per Thread
> - Game ID

Agree. I thought we would do this already in a slightly less sophisticated way for generating the UUIDs for the Amphora output objects for an ephemeral program execution. However, looking at the source code, it looks like as we currently support a single output object per execution only.

https://github.com/carbynestack/ephemeral/blob/70795c1e688fdffef3761431c341f1302f0105b0/pkg/ephemeral/io/feeder.go#L140-L160

Created an issue to track that #13.

> Would such a multi-thread approach create special requirements to the pod running ephemeral (like how many vCores it has or so?)

Go uses cooperative scheduling of goroutines, so I don't think there is a requirement on that side. However, to leverage the potential speed up when using MP-SPDZ threads, the Knative service should be configured with the right resource limits (see https://knative.dev/docs/serving/services/configure-requests-limits-services/#additional-resources)

**Comment by sbckr** [comment]
Resolved by #27
